src/Microsoft.OData.Core/ODataReaderCoreAsync.cs

```diff
@@ -175,95 +177,206 @@ protected virtual Task<bool> ReadAtDeltaDeletedLinkImplementationAsync()
-                    result = TaskUtils.GetTaskForSynchronousOperation(() => this.IncreaseResourceDepth())
-                        .FollowOnSuccessWithTask(t => this.ReadAtDeletedResourceStartImplementationAsync());
+                    this.IncreaseResourceDepth();
+                    result = await this.ReadAtDeletedResourceStartImplementationAsync()
+                        .ConfigureAwait(false);
                     break;
 
                 case ODataReaderState.DeletedResourceEnd:
-                    result = TaskUtils.GetTaskForSynchronousOperation(() => this.DecreaseResourceDepth())
-                        .FollowOnSuccessWithTask(t => this.ReadAtDeletedResourceEndImplementationAsync());
+                    this.DecreaseResourceDepth();
+                    result = await this.ReadAtDeletedResourceEndImplementationAsync()
+                        .ConfigureAwait(false);
                     break;
 
                 case ODataReaderState.DeltaLink:
-                    result = this.ReadAtDeltaLinkImplementationAsync();
+                    result = await this.ReadAtDeltaLinkImplementationAsync()
+                        .ConfigureAwait(false);
                     break;
 
                 case ODataReaderState.DeltaDeletedLink:
-                    result = this.ReadAtDeltaDeletedLinkImplementationAsync();
+                    result = await this.ReadAtDeltaDeletedLinkImplementationAsync()
+                        .ConfigureAwait(false);
                     break;
 
                 case ODataReaderState.Exception:    // fall through
                 case ODataReaderState.Completed:
-                    result = TaskUtils.GetFaultedTask<bool>(new ODataException(Strings.ODataReaderCore_NoReadCallsAllowed(this.State)));
-                    break;
+                    throw new ODataException(Strings.ODataReaderCore_NoReadCallsAllowed(this.State));
 
                 default:
                     Debug.Assert(false, "Unsupported reader state " + this.State + " detected.");
-                    result = TaskUtils.GetFaultedTask<bool>(new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataReaderCoreAsync_ReadAsynchronously)));
-                    break;
+                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataReaderCoreAsync_ReadAsynchronously));
             }
 
-            return result.FollowOnSuccessWith(t => t.Result);
+            return result;
+        }
+
+        /// <summary>
+        /// Asynchronously creates a <see cref="Stream"/> for reading a stream property
+        /// when reader in state <see cref="ODataReaderState.Stream"/>.
+        /// </summary>
+        /// <returns>
+        /// A task that represents the asynchronous operation.
+        /// The value of the TResult parameter contains a <see cref="Stream"/> for reading the stream property.
+        /// </returns>
+        protected virtual Task<Stream> CreateReadStreamImplementationAsync()

```

Comment:
make all as abstract methods?

don't we add 'async' identifier?

Reply:
@KenitoInc For most of those methods you see declared as `public virtual` (especially in public classes or interfaces), its because they were introduced midway and it done that way to prevent a breaking change. The safest time to make those methods `abstract` is in the next major release.

Reply:
@xuzhg We don't need to add `async` identifier for those `protected virtual` methods. The only purpose the `throw new NotImplementedException();` stub serves is to not force all classes that inherit from this class implement those new methods if there's no need to. `async` identifier or no, it'd be a bad day if any of the derived classes was to ship before a proper implementation for those methods has been supplied and any of those methods got called.

Reply:
About making the methods `abstract`, in as much as this class is `internal`, I'm inclined to follow what appears to be the established practice going by the existing code - new methods in `abstract` classes are introduced as `virtual`.